taichi/analysis/same_statements.cpp

```diff
@@ -13,23 +14,35 @@ class IRNodeComparator : public IRVisitor {
   IRNode *other_node;
   // map the id from this node to the other node
   std::unordered_map<int, int> id_map;
+
   // ids which don't belong to either node
   std::unordered_set<int> captured_id;
+  bool implicitly_capture_ids;
 
  public:
   bool same;
 
-  explicit IRNodeComparator(IRNode *other_node) : other_node(other_node) {
+  explicit IRNodeComparator(IRNode *other_node,
+                            std::optional<std::unordered_map<int, int>> id_map)
+      : other_node(other_node) {
     allow_undefined_visitor = true;
     invoke_default_visitor = true;
     same = true;
+    if (id_map.has_value()) {
+      implicitly_capture_ids = false;
+      this->id_map = std::move(id_map.value());
+    } else {
+      implicitly_capture_ids = true;
+    }
   }
 
   void map_id(int this_id, int other_id) {
-    if (captured_id.find(this_id) != captured_id.end() ||
-        captured_id.find(other_id) != captured_id.end()) {
-      same = false;
-      return;
+    if (implicitly_capture_ids) {
+      if (captured_id.find(this_id) != captured_id.end() ||
+          captured_id.find(other_id) != captured_id.end()) {
```

Comment:
Add a line explain this condition?

Reply:
I think this condition can never happen in one kernel now (at least if `re_id` is called before)...
Suppose `same_statements(x, y)` is called (`x` and `y` are `Stmt::id`s), when we are going to map `a` to `b`, we have `a >= x` and `b >= y`. If some id `c` is captured before we map `a` to `b`, we have `c < a` and `c < b` because the id of a statement's operands must be smaller than the id of that statement.

Reply:
Simplified in https://github.com/taichi-dev/taichi/pull/2012/commits/adc1518de957ce799b9c20b5621ba693735b3772 if this condition does cannot happen.